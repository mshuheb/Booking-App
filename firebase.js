import { getApp, initializeApp } from "firebase/app";
import { initializeAuth, getReactNativePersistence } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import ReactNativeAsyncStorage from "@react-native-async-storage/async-storage";

const firebaseConfig = {
  apiKey: "AIzaSyBRqeTVnFdloEn3bpB9fKk_btNd0tPKWok",
  authDomain: "booking-project-afa36.firebaseapp.com",
  projectId: "booking-project-afa36",
  storageBucket: "booking-project-afa36.appspot.com",
  messagingSenderId: "1096229966164",
  appId: "1:1096229966164:web:c1238ddd012239251e38c1",
};

let app;
try {
  app = getApp();
} catch {
  app = initializeApp(firebaseConfig);
}

const auth = initializeAuth(app, {
  persistence: getReactNativePersistence(ReactNativeAsyncStorage),
});

const db = getFirestore();

export { auth, db };
