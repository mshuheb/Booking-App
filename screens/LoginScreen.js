import {
  SafeAreaView,
  KeyboardAvoidingView,
  TextInput,
  Pressable,
  Text,
  View,
  ActivityIndicator,
} from "react-native";
import React, { useState } from "react";
import { useNavigation } from "@react-navigation/native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { auth } from "../firebase";
import { signInWithEmailAndPassword } from "firebase/auth";

const LoginScreen = () => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [isLoggingIn, setIsLoggingIn] = useState(false);
  const navigation = useNavigation();

  const login = async () => {
    setIsLoggingIn(true);
    try {
      const userCredential = await signInWithEmailAndPassword(auth, email, password);
      const user = userCredential.user;
      const token = await user.getIdToken();

      await AsyncStorage.setItem("tokenUser", token);
      await AsyncStorage.setItem("userEmail", user.email);
      const phoneNumber = user.phoneNumber || "";
      await AsyncStorage.setItem("userPhoneNumber", phoneNumber);

      console.log("Login successful!");
      navigation.replace("Main");
    } catch (error) {
      console.error("Login failed:", error.message);
      alert("Login failed. Please check your email and password.");
    } finally {
      setIsLoggingIn(false);
    }
  };

  return (
    <SafeAreaView
      style={{
        flex: 1,
        backgroundColor: "white",
        padding: 10,
        alignItems: "center",
      }}
    >
      <KeyboardAvoidingView>
        <View style={{ justifyContent: "center", alignItems: "center", marginTop: 100 }}>
          <Text style={{ color: "#003580", fontSize: 17, fontWeight: "700" }}>
            Sign In
          </Text>
          <Text style={{ marginTop: 15, fontSize: 18, fontWeight: "500" }}>
            Sign In to Your Account
          </Text>
        </View>

        <View style={{ marginTop: 50 }}>
          <View>
            <Text style={{ fontSize: 18, fontWeight: "600", color: "gray" }}>
              Email
            </Text>

            <TextInput
              value={email}
              onChangeText={setEmail}
              placeholder="Enter your email id"
              placeholderTextColor={"black"}
              style={{
                fontSize: 18,
                borderBottomColor: "gray",
                borderBottomWidth: 1,
                marginVertical: 10,
                width: 300,
              }}
            />
          </View>

          <View style={{ marginTop: 15 }}>
            <Text style={{ fontSize: 18, fontWeight: "600", color: "gray" }}>
              Password
            </Text>

            <TextInput
              value={password}
              onChangeText={setPassword}
              secureTextEntry={true}
              placeholder="Password"
              placeholderTextColor={"black"}
              style={{
                fontSize: 18,
                borderBottomColor: "gray",
                borderBottomWidth: 1,
                marginVertical: 10,
                width: 300,
              }}
            />
          </View>
        </View>

        {isLoggingIn ? (
          <ActivityIndicator size="large" color="#003580" style={{ marginTop: 50 }} testID="ActivityIndicator" />
        ) : (
          <Pressable
            onPress={login}
            style={{
              width: 200,
              backgroundColor: "#003580",
              padding: 15,
              borderRadius: 7,
              marginTop: 50,
              marginLeft: "auto",
              marginRight: "auto",
            }}
          >
            <Text style={{ textAlign: "center", color: "white", fontSize: 17, fontWeight: "bold" }}>
              Login
            </Text>
          </Pressable>
        )}

        <Pressable onPress={() => navigation.navigate("Register")} style={{ marginTop: 20 }}>
          <Text style={{ textAlign: "center", color: "gray", fontSize: 17 }}>
            Don't have an account? Sign up
          </Text>
        </Pressable>
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

export default LoginScreen;
