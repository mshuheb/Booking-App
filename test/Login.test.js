import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { Alert } from "react-native";
import LoginScreen from "../screens/LoginScreen";
import { signInWithEmailAndPassword } from "firebase/auth";
import { useNavigation } from "@react-navigation/native";


jest.mock("firebase/auth", () => {
  return {
    signInWithEmailAndPassword: jest.fn(),
    initializeAuth: jest.fn(),
    getReactNativePersistence: jest.fn(), 
  };
});

jest.mock("@react-native-async-storage/async-storage", () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
}));

jest.mock("@react-navigation/native", () => ({
  useNavigation: jest.fn(),
}));

jest.spyOn(Alert, "alert").mockImplementation(() => {});

describe("<LoginScreen />", () => {
  let navigationMock;

  beforeEach(() => {
    navigationMock = { replace: jest.fn(), navigate: jest.fn() };
    useNavigation.mockReturnValue(navigationMock);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders correctly with initial UI elements", () => {
    const { getByPlaceholderText, getByText } = render(<LoginScreen />);

    expect(getByText("Sign In")).toBeTruthy();
    expect(getByText("Sign In to Your Account")).toBeTruthy();
    expect(getByPlaceholderText("Enter your email id")).toBeTruthy();
    expect(getByPlaceholderText("Password")).toBeTruthy();
    expect(getByText("Login")).toBeTruthy();
  });

  test("matches snapshot", () => {
    const { toJSON } = render(<LoginScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  test("allows user to input email and password", () => {
    const { getByPlaceholderText } = render(<LoginScreen />);

    const emailInput = getByPlaceholderText("Enter your email id");
    const passwordInput = getByPlaceholderText("Password");

    fireEvent.changeText(emailInput, "test@example.com");
    fireEvent.changeText(passwordInput, "password123");

    expect(emailInput.props.value).toBe("test@example.com");
    expect(passwordInput.props.value).toBe("password123");
  });

  test("shows activity indicator during login", async () => {
    signInWithEmailAndPassword.mockImplementation(() => new Promise(() => {})); 

    const { getByText, getByPlaceholderText, getByTestId } = render(<LoginScreen />);

    const emailInput = getByPlaceholderText("Enter your email id");
    const passwordInput = getByPlaceholderText("Password");
    const loginButton = getByText("Login");

    fireEvent.changeText(emailInput, "test@example.com");
    fireEvent.changeText(passwordInput, "password123");
    fireEvent.press(loginButton);

    await waitFor(() => {
      expect(getByTestId("ActivityIndicator")).toBeTruthy();
    });
  });

  test("redirects to register screen on 'Sign up' press", () => {
    const { getByText } = render(<LoginScreen />);
    const signUpText = getByText("Don't have an account? Sign up");

    fireEvent.press(signUpText);

    expect(navigationMock.navigate).toHaveBeenCalledWith("Register");
  });
});
