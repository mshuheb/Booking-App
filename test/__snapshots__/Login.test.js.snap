// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 100,
        }
      }
    >
      <Text
        style={
          {
            "color": "#003580",
            "fontSize": 17,
            "fontWeight": "700",
          }
        }
      >
        Sign In
      </Text>
      <Text
        style={
          {
            "fontSize": 18,
            "fontWeight": "500",
            "marginTop": 15,
          }
        }
      >
        Sign In to Your Account
      </Text>
    </View>
    <View
      style={
        {
          "marginTop": 50,
        }
      }
    >
      <View>
        <Text
          style={
            {
              "color": "gray",
              "fontSize": 18,
              "fontWeight": "600",
            }
          }
        >
          Email
        </Text>
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter your email id"
          placeholderTextColor="black"
          style={
            {
              "borderBottomColor": "gray",
              "borderBottomWidth": 1,
              "fontSize": 18,
              "marginVertical": 10,
              "width": 300,
            }
          }
          value=""
        />
      </View>
      <View
        style={
          {
            "marginTop": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "gray",
              "fontSize": 18,
              "fontWeight": "600",
            }
          }
        >
          Password
        </Text>
        <TextInput
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="black"
          secureTextEntry={true}
          style={
            {
              "borderBottomColor": "gray",
              "borderBottomWidth": 1,
              "fontSize": 18,
              "marginVertical": 10,
              "width": 300,
            }
          }
          value=""
        />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#003580",
          "borderRadius": 7,
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginTop": 50,
          "padding": 15,
          "width": 200,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 17,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Login
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          {
            "color": "gray",
            "fontSize": 17,
            "textAlign": "center",
          }
        }
      >
        Don't have an account? Sign up
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
